<br>

$\LARGE \textsf{PS1 - Modelling & Optimization}$

<br>

---

$\large \textsf{EXERCISE 4.}$  

Consider the linear programming problem in the standard maximization form, with 3 variables and 3 constraints, denoted by the following matrices:

$$
A = \begin{bmatrix}
0 & 1 & 5 \\
1 & 2 & 0 \\
1 & 1 & -2 \\
\end{bmatrix},
\quad
b = \begin{bmatrix}
5 \\
2 \\
1 \\
\end{bmatrix},
\quad
c = \begin{bmatrix}
3 & 4 & 10 \\
\end{bmatrix}
$$

The following simplex tableau is associated with the optimal solution for the problem. Answer every question in regard to this tableau/solution, showing all your computations.

| Max || x1 | x2 | x3 | s1 | s2 | s3 |  
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |  
| z | 16 | 0 | 4 | 0 | 2 | 3 | 0 |  
| x3 | 1 | 0 | $\frac{1}{5}$ | 1 | $\frac{1}{5}$ | 0 | 0 |  
| x1 | 2 | 1 | 2 | 0 | 0 | 1 | 0 |  
| s3 | 1 | 0 | $\frac{-3}{5}$ | 0 | $\frac{2}{5}$ | -1 | 1 |  

---

<br> 

> $\small \textsf{PART A.}$ Show that the basic solution associated to this tableau is optimal, indicating the basic and non-basic variables, their respective values and the optimal value for the problem. Write the $B$ and $B ^ {−1}$ matrices.

In [1]:
# B1, i.e. the first Basis Matrix, given by original tableau's columns for x3, x1, and s3, respectively
B1 = [5 0 0; 0 1 0; -2 1 1]

3×3 Array{Int64,2}:
  5  0  0
  0  1  0
 -2  1  1

In [2]:
# The inverse of this basis matrix is indicated by columns for s1, s2, and s3 for rows 1, 2, and 3
B1_inv = [1/5 0 0; 0 1 0; 2/5 -1 1]

3×3 Array{Float64,2}:
 0.2   0.0  0.0
 0.0   1.0  0.0
 0.4  -1.0  1.0

In [3]:
# Note that:
B1_inv == inv(B1)

true

Given that this is a maximization problem and that all reduced costs are non-negative and the RHS is non-negative, the basic solution is, respectively, optimal and feasible.

Therefore, the basic feasible solution is given by:

$$ \textrm{BFS} = (2, 0, 1, 0, 0, 1) \ \textrm{where} \ z = 16 $$

In [60]:
# z0 is given by y*b = c_bv * Bˆ{-1} * b
y1 = [10 3 0] * B1_inv
b1 = [5 2 1]'
z1 = y1 * b1

1×1 Array{Float64,2}:
 16.0

Thus:

$$
B_1 = \begin{bmatrix}
5 & 0 & 0 \\
0 & 1 & 0 \\
-2 & 1 & 1 \\
\end{bmatrix},
\quad
B_1 ^ {-1} = \begin{bmatrix}
\frac{1}{5} & 0 & 0 \\
0 & 1 & 0 \\
\frac{2}{5} & -1 & 1 \\
\end{bmatrix}
$$

<br>

---

<br> 

> $\small \textsf{PART B.}$ Write the corresponding dual problem and indicate its optimal solution based on the given information (do not solve the dual problem). Show that the Complementary Conditions are satisfied.

<br>

---

<br> 

> $\small \textsf{PART C.}$ Consider a change in the right hand side of the second constraint from $2$ to $4$. How does this affect the current solution? If pertinent, find the new optimal value and/or solution.

Changes to the right-hand side do not affect row zero of the tableau.

They may, however, affect the feasibility of the basic feasible solution. To understand whether this is the case or not, we want to get the new RHS after performing EROs given by the $B_1 ^{-1}$ matrix.



In [6]:
b2 = [5 4 1]'

3×1 Array{Int64,2}:
 5
 4
 1

In [9]:
z2 = y1 * b2

1×1 Array{Float64,2}:
 22.0

At this point, we can see that $z_2$ has increased and is now 'super-optimal'. Once primal feasibility is achieved, we will get the optimal $z$ value.

Essentially:

$$
B_1 ^ {-1} \overline{b}
=
\begin{bmatrix}
\frac{1}{5} & 0 & 0 \\
0 & 1 & 0 \\
\frac{2}{5} & -1 & 1 \\
\end{bmatrix}
\times
\begin{bmatrix}
5 \\
4 \\
1 \\
\end{bmatrix}
=
\begin{bmatrix}
1 \\
4 \\
-1 \\
\end{bmatrix}
\quad \Longrightarrow \quad
\textit{primal basic solution is no longer feasible}
$$

In [10]:
RHS0 = B1_inv * b1
RHS1 = B1_inv * b2

3×1 Array{Float64,2}:
  1.0
  4.0
 -1.0

We do not have primal feasibility, as mentioned previously. We must perform the dual simplex method in order to regain feasibility. We know at once that the leaving variable will be third element of $\textit{basis}_1 = \{ x_3, x_1, s_3 \}$ - i.e., $s_3$ leaves $\textit{basis}_1$.

But what variable now enters the basis?

| Max || x1 | x2 | x3 | s1 | s2 | s3 |  
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |  
| z | 16 | 0 | 4 | 0 | 2 | 3 | 0 |  
| x3 | 1 | 0 | $\frac{1}{5}$ | 1 | $\frac{1}{5}$ | 0 | 0 |  
| x1 | 4 | 1 | 2 | 0 | 0 | 1 | 0 |  
| s3 | -1 | 0 | $\frac{-3}{5}$ | 0 | $\frac{2}{5}$ | -1 | 1 | 

In [11]:
# Defining the original A matrix
A = [0 1 5; 1 2 0; 1 1 -2]

3×3 Array{Int64,2}:
 0  1   5
 1  2   0
 1  1  -2

In [23]:
# The original tableau
tb1 = [A eye(3,3)]

3×6 Array{Float64,2}:
 0.0  1.0   5.0  1.0  0.0  0.0
 1.0  2.0   0.0  0.0  1.0  0.0
 1.0  1.0  -2.0  0.0  0.0  1.0

In [24]:
# The current tableau
tb2 = B1_inv * tb0

3×6 Array{Float64,2}:
 0.0   0.2  1.0  0.2   0.0  0.0
 1.0   2.0  0.0  0.0   1.0  0.0
 0.0  -0.6  0.0  0.4  -1.0  1.0

In [29]:
# Row 0 for coefficients in the original tableau
all_coeff_1 = [3 4 10 0 0 0]

# Row 0 for coefficients in the current tableau
all_coeff_2 = y1 * tb1 - all_coeff_1

1×6 Array{Float64,2}:
 0.0  4.0  0.0  2.0  3.0  0.0

In [61]:
# Creating a list with results of min ratio tests for dual simplex method
min_ratio_dual = ["#" "x1" "x2" "x3" "s1" "s2" "s3" ]
iter = 1
min_ratio_dual = [min_ratio_dual; iter all_coeff_2 ./ tb2[3,:]']

2×7 Array{Any,2}:
  "#"     "x1"    "x2"       "x3"   "s1"    "s2"   "s3"
 1.0   NaN      -6.66667  NaN      5.0    -3.0    0.0  

The entering variable is $s_2$, as for the dual min ratio test, the respective column values for the row of the exiting variable must be negative. Out of the set of these negtive values, we look for the minimum - in this case, $-0.15$.

As such:

In [45]:
B1_inv * A[:,2]

3-element Array{Float64,1}:
  0.2
  2.0
 -0.6

$$
B_1 ^ {-1}
\times
A ^ {x_2}
=
\begin{bmatrix}
\frac{1}{5} \\
2 \\
\frac{-3}{5} \\
\end{bmatrix}
= \begin{bmatrix}
a_{12} \\
a_{22} \\
a_{32} \\
\end{bmatrix}
\quad \Longrightarrow \quad
\textrm{Column for} \ x_2 \ \textrm{in matrix} \ A
$$

$$
\begin{aligned}
B_2 & = \begin{bmatrix}
5 & 0 & 1 \\
0 & 1 & 2 \\
-2 & 1 & 1 \\
\end{bmatrix},
\quad
B_2 ^ {-1} = E_1 B_1 ^ {-1}
& =
\begin{bmatrix}
1 &  0 & \frac{-a_{12}}{a_{32}} \\
0 & 1  & \frac{-a_{22}}{a_{32}} \\
0 &  0 & \frac{1}{a_{32}} \\
\end{bmatrix}
\times
\begin{bmatrix}
\frac{1}{5} & 0 & 0 \\
0 & 1 & 0 \\
\frac{2}{5} & -1 & 1 \\
\end{bmatrix}
\\
\\
&& =
\begin{bmatrix}
1 &  0 & \frac{1}{3} \\
0 & 1  & \frac{10}{3} \\
0 &  0 & \frac{-5}{3} \\
\end{bmatrix}
\times
\begin{bmatrix}
\frac{1}{5} & 0 & 0 \\
0 & 1 & 0 \\
\frac{2}{5} & -1 & 1 \\
\end{bmatrix}
\end{aligned}
$$

<br>

As such:

$$
B_2  =
\begin{bmatrix}
5 & 0 & 1 \\
0 & 1 & 2 \\
-2 & 1 & 1 \\
\end{bmatrix},
\quad
\textrm{with}
\quad
B_2 ^ {-1}
 = 
\begin{bmatrix}
\frac{1}{3} & \frac{-1}{3} & \frac{1}{3} \\
\frac{4}{3} & \frac{-7}{3} & \frac{10}{3} \\
\frac{-2}{3} & \frac{5}{3} & \frac{-5}{3} \\
\end{bmatrix}
$$

In [46]:
E1 = [1 0 1/3; 0 1 10/3; 0 0 -5/3]

3×3 Array{Float64,2}:
 1.0  0.0   0.333333
 0.0  1.0   3.33333 
 0.0  0.0  -1.66667 

In [47]:
B2 = [5 0 1; 0 1 2; -2 1 1]

3×3 Array{Int64,2}:
  5  0  1
  0  1  2
 -2  1  1

In [49]:
B2_inv = E1 * B1_inv

3×3 Array{Float64,2}:
  0.333333  -0.333333   0.333333
  1.33333   -2.33333    3.33333 
 -0.666667   1.66667   -1.66667 

<br>

We must get the new values for the new tableau:

In [53]:
y2 = [10 3 4] * B2_inv

1×3 Array{Float64,2}:
 4.66667  -3.66667  6.66667

In [54]:
tb3 = B2_inv * tb0

3×6 Array{Float64,2}:
 0.0  5.55112e-17   1.0           0.333333  -0.333333   0.333333
 1.0  0.0           8.88178e-16   1.33333   -2.33333    3.33333 
 0.0  1.0          -4.44089e-16  -0.666667   1.66667   -1.66667 

In [55]:
# Row 0 for coefficients in the current tableau
all_coeff_3 = y2 * tb1 - all_coeff_1

1×6 Array{Float64,2}:
 0.0  1.77636e-15  5.32907e-15  4.66667  -3.66667  6.66667

In [58]:
# The new RHS is:
RHS2 = B2_inv * b2

3×1 Array{Float64,2}:
 0.666667
 0.666667
 1.66667 

In [59]:
# The new optimal value is:
z3 = y2*b2

1×1 Array{Float64,2}:
 15.3333

Therefore, the current basic feasible solution is given by:

$$ \textrm{BFS} = \left ( \frac{2}{3}, \frac{5}{3}, \frac{2}{3}, 0, 0, 0 \right ) \ \textrm{where} \ z = 15.33 $$

<br>